src/RefractDataStructure.cc

```diff
@@ -446,55 +405,97 @@ namespace drafter {
+                data.enumerations.push_back(Fetch<E>()(typeNames, context));
+            }
+        };
+
+        ExtractTypeDefinition(ElementData<ElementType>& data, const ConversionContext& context) : data(data), context(context) {}
+
+        void operator()(const NodeInfo<mson::TypeDefinition>& typeDefinition) {
+            Store<ElementType>()(data, typeDefinition.node->typeSpecification.nestedTypes, context);
         }
     };
 
+    /**
+     * Extract mson definitions like
+     * - value: vm1, vm2
+     */
     template <typename T, typename V = typename T::ValueType>
     struct ExtractValueMember
     {
-        typedef T ElementType;
-        typedef typename ElementData<T>::ValueInfo ValueInfo;
+        using ElementType = T;
+        using ElementInfo = typename ElementData<T>::ElementInfo;
 
         ElementData<T>& data;
         ConversionContext& context;
 
-        template <typename U, bool dummy = true>
-        struct Fetch {  // primitive values
+        template <typename U, typename IsPrimitive = typename ElementData<U>::IsPrimitive, bool dummy = true> struct Fetch;
 
-            ValueInfo operator()(const NodeInfo<mson::ValueMember>& valueMember, ConversionContext& context) {
+        template <typename U, bool dummy>
+        struct Fetch <U, PrimitiveType, dummy> {  // primitive values
+
+            ElementInfo operator()(const NodeInfo<mson::ValueMember>& valueMember, ConversionContext& context) {
                 if (valueMember.node->valueDefinition.values.size() > 1) {
                     throw snowcrash::Error("only one value is supported for primitive types", snowcrash::MSONError, valueMember.sourceMap->sourceMap);
                 }
 
                 const mson::Value& value = *valueMember.node->valueDefinition.values.begin();
+                snowcrash::SourceMap<V> sourceMap = FetchSourceMap<V>()(valueMember);
+
+                return std::make_tuple(value.literal, sourceMap);
+
+            }
+        };
+
+        template<typename U, bool dummy>
+        struct Fetch<U, ComplexType, dummy> { // Array|Object
+
+            ElementInfo operator()(const NodeInfo<mson::ValueMember>& valueMember, ConversionContext& context) {
 
-                std::pair<bool, U> val = LiteralTo<U>(value.literal);
-                snowcrash::SourceMap<U> sourceMap = FetchSourceMap<U>()(valueMember);
+                const mson::BaseTypeName type = SelectNestedTypeSpecification(valueMember.node->valueDefinition.typeDefinition.typeSpecification.nestedTypes);
+
+                const RefractElementFactory& f = FactoryFromType(type);
+                const mson::Values& values = valueMember.node->valueDefinition.values;
 
-                ValueInfo result = std::make_tuple(val.second, sourceMap, val.first);
+                RefractElements elements;
 
-                CheckValueValidity<T>()(result, context);
+                for (mson::Values::const_iterator it = values.begin(); it != values.end(); ++it) {
```

Comment:
Can we simplify this with high order functions like map while reducing the amount of mutation (and possibly allowing the optimiser to do a better job)?

```cpp
const RefractElements elements = std::transform(values.begin(), values.end(), ^ (mson::Values::const_iterator iterator) {
    return f.Create(it->literal, it->variable ? eSample : eValue);
});
```

I apologise, this example code is Pseudo-C++ as I do not remember the C++ block syntax off the top of my head